addon/components/flash-message.js

```diff
@@ -2,127 +2,110 @@ import { htmlSafe, classify } from '@ember/string';
-  },
-
-  willDestroyElement() {
-    this._super(...arguments);
-    this.element.removeEventListener('mouseenter', this._mouseEnterHandler);
-    this.element.removeEventListener('mouseleave', this._mouseLeaveHandler);
-  },
-
-  progressDuration: computed('flash.showProgress', {
-    get() {
-      if (!get(this, 'flash.showProgress')) {
-        return false;
-      }
-
-      const duration = getWithDefault(this, 'flash.timeout', 0);
-
-      return htmlSafe(`transition-duration: ${duration}ms`);
+  @readOnly('flash.showProgress')
+  showProgress;
+
+  @not('exiting')
+  notExiting;
+
+  @and('showProgress', 'notExiting')
+  showProgressBar;
+
+  @readOnly('flash.exiting')
+  exiting;
+
+  @bool('template')
+  hasBlock;
+
+  @computed('flash.type')
+  get alertType() {
+    const flashType = this.flash.type || '';
+    const messageStyle = this.messageStyle || '';
+    let prefix = 'alert alert-';
+
+    if (messageStyle === 'foundation') {
+      prefix = 'alert-box ';
     }
-  }),
 
-  click() {
-    const destroyOnClick = getWithDefault(this, 'flash.destroyOnClick', true);
+    return `${prefix}${flashType}`;
+  }
 
-    if (destroyOnClick) {
-      this._destroyFlashMessage();
+  @computed('flash.type')
+  get flashType() {
+    return classify(this.flash.type || '');
+  }
+
+  @computed('flash.showProgress')
+  get progressDuration() {
+    if (!this.flash?.showProgress) {
+      return false;
     }
-  },
+    const duration = this.flash?.timeout || 0;
+    return htmlSafe(`transition-duration: ${duration}ms`);
+  }
 
   _mouseEnter() {
-    const flash = get(this, 'flash');
-    if (isPresent(flash)) {
-      flash.preventExit();
+    if (isPresent(this.flash)) {
+      this.flash.preventExit();
     }
-  },
+  }
 
   _mouseLeave() {
-    const flash = get(this, 'flash');
-    if (isPresent(flash) && !get(flash, 'exiting')) {
-      flash.allowExit();
+    if (isPresent(this.flash) && !this.flash.exiting) {
+      this.flash.allowExit();
     }
-  },
-
-  willDestroy() {
-    this._super(...arguments);
-    this._destroyFlashMessage();
-    cancel(get(this, 'pendingSet'));
-  },
+  }
 
-  // private
   _destroyFlashMessage() {
-    const flash = getWithDefault(this, 'flash', false);
-
-    if (flash) {
-      flash.destroyMessage();
+    if (this.flash) {
+      this.flash.destroyMessage();
     }
-  },
+  }
 
-  actions: {
-    close() {
+  @action
+  onClick() {
+    const destroyOnClick = this.flash?.destroyOnClick ?? true;
+
+    if (destroyOnClick) {
       this._destroyFlashMessage();
     }
   }
-});
+
+  @action
+  onClose() {
+    this._destroyFlashMessage();
+  }
+
+  @action
+  onDidInsert(element) {
+    const pendingSet = next(this, () => {
+      set(this, 'active', true);
+    });
+    set(this, 'pendingSet', pendingSet);
+    set(this, '_mouseEnterHandler', this._mouseEnter.bind(this));
+    set(this, '_mouseLeaveHandler', this._mouseLeave.bind(this));
```

Comment:
Is there an advantage to doing this vs. using the `@action` decorator?

Reply:
This function does use the `@action` decorator and is called on "init" via the `did-insert` helper in the template.

Reply:
Should be able to get rid of the `.bind` stuff then. `@action` does this for you!

Reply:
Good call 👍